### PR TITLE
(require|disallow)NewlineBeforeBlockStatements: Fine-grained control

### PR DESCRIPTION
**Pull request by davejhilton**
For the `requireNewlineBeforeBlockStatements` and `disallowNewlineBeforeBlockStatements` rules, this adds the ability to specify an array of block statement types that it should enforce, rather than merely having a binary switch via specifying `true` to enforce it for _all_ block statements, or omitting the rule to enforce it for _none_ of them.

The use-case this targets is one where my team's code style conventions dictate that we keep curly braces on the same line for functions, but on a new line for other control statements (`if`, `else`, `for`, `while`, etc).

e.g.,
```js
function getAsyncValue(callback) {     // brace on same line for functions
    setTimeout(function() {            // brace on same line for functions
        callback(Math.Floor(Math.random() * 10000));
    });
}

...

getAsyncValue(function(value) {        // brace on same line for functions
    var result = 0;
    var size;

    if (value > 1024)                  // brace on new line for other control statements
    {
        result = largeNumberAggregator(yValue);
        size = 'large';
    }
    else                               // brace on new line for other control statements
    {
        result = smallNumberAggregator(yValue);
        size = 'small';
    }

    console.log('Crunched a ' + size + ' number (' + value + ') and got ' + result);
});
```

could be style-checked by using the following rules:
```js
{
    ...
    "requireNewlineBeforeBlockStatements": [
        "if",
        "else",
        "for",
        "while",
        "do",
        "try",
        "catch",
        "finally"
    ],
    "disallowNewlineBeforeBlockStatements": ["function"],
    ...
}
```

### CODE REVIEW

File: lib/rules/disallow-newline-before-block-statements.js [hunk 1]
```diff
@@ -90,15 +184,82 @@ module.exports.prototype = {
...
+                blockType = 'finally';
+            } else {
+                blockType = 'try';
+            }
+            break;
+        case 'CatchClause':
+            blockType = 'catch';
+            break;
+        default:
+            blockType = 'other';
```
Comment (mikesherov) [inline lib/rules/disallow-newline-before-block-statements.js]: just return here.

File: lib/rules/disallow-newline-before-block-statements.js [hunk 2]
```diff
@@ -90,15 +184,82 @@ module.exports.prototype = {
...
+            break;
+        case 'TryStatement':
+            if (node.parentNode.finalizer === node) {
+                blockType = 'finally';
+            } else {
+                blockType = 'try';
+            }
+            break;
+        case 'CatchClause':
+            blockType = 'catch';
```
Comment (mikesherov) [inline lib/rules/disallow-newline-before-block-statements.js]: just return here.

File: lib/rules/disallow-newline-before-block-statements.js [hunk 3]
```diff
@@ -90,15 +184,82 @@ module.exports.prototype = {
...
+            blockType = 'for';
+            break;
+        case 'WhileStatement':
+            blockType = 'while';
+            break;
+        case 'DoWhileStatement':
+            blockType = 'do';
+            break;
+        case 'TryStatement':
+            if (node.parentNode.finalizer === node) {
```
Comment (mikesherov) [inline lib/rules/disallow-newline-before-block-statements.js]: this whole block is: `return node.parentNode.finalizer === node ? 'finally' : 'try'`

File: lib/rules/disallow-newline-before-block-statements.js [hunk 4]
```diff
@@ -90,15 +184,82 @@ module.exports.prototype = {
...
+        case 'ForStatement':
+        case 'ForInStatement':
+        case 'ForOfStatement':
+            blockType = 'for';
+            break;
+        case 'WhileStatement':
+            blockType = 'while';
+            break;
+        case 'DoWhileStatement':
+            blockType = 'do';
```
Comment (mikesherov) [inline lib/rules/disallow-newline-before-block-statements.js]: just return here.

File: lib/rules/disallow-newline-before-block-statements.js [hunk 5]
```diff
@@ -90,15 +184,82 @@ module.exports.prototype = {
...
+        case 'FunctionExpression':
+            blockType = 'function';
+            break;
+        case 'ForStatement':
+        case 'ForInStatement':
+        case 'ForOfStatement':
+            blockType = 'for';
+            break;
+        case 'WhileStatement':
+            blockType = 'while';
```
Comment (mikesherov) [inline lib/rules/disallow-newline-before-block-statements.js]: just return here.

File: lib/rules/disallow-newline-before-block-statements.js [hunk 6]
```diff
@@ -90,15 +184,82 @@ module.exports.prototype = {
...
+            }
+            break;
+        case 'FunctionDeclaration':
+        case 'FunctionExpression':
+            blockType = 'function';
+            break;
+        case 'ForStatement':
+        case 'ForInStatement':
+        case 'ForOfStatement':
+            blockType = 'for';
```
Comment (mikesherov) [inline lib/rules/disallow-newline-before-block-statements.js]: just return here.

File: lib/rules/disallow-newline-before-block-statements.js [hunk 7]
```diff
@@ -90,15 +184,82 @@ module.exports.prototype = {
...
+        case 'IfStatement':
+            if (node.parentNode.alternate === node) {
+                blockType = 'else';
+            } else {
+                blockType = 'if';
+            }
+            break;
+        case 'FunctionDeclaration':
+        case 'FunctionExpression':
+            blockType = 'function';
```
Comment (mikesherov) [inline lib/rules/disallow-newline-before-block-statements.js]: just return here.

File: lib/rules/disallow-newline-before-block-statements.js [hunk 8]
```diff
@@ -90,15 +184,82 @@ module.exports.prototype = {
...
+        }
+    });
+}
+
+function getBlockType(node) {
+    var parentType = node.parentNode.type;
+    var blockType;
+    switch (parentType) {
+        case 'IfStatement':
+            if (node.parentNode.alternate === node) {
```
Comment (mikesherov) [inline lib/rules/disallow-newline-before-block-statements.js]: This whole block is: `return node.parentNode.alternate === node ? 'else' : 'if';`

File: lib/rules/disallow-newline-before-block-statements.js [hunk 9]
```diff
@@ -90,15 +184,82 @@ module.exports.prototype = {
...
+                message: 'Newline before curly brace for "' + type + '" block statement is disallowed'
             });
-        });
+        }
+    });
+}
+
+function getBlockType(node) {
+    var parentType = node.parentNode.type;
+    var blockType;
```
Comment (mikesherov) [inline lib/rules/disallow-newline-before-block-statements.js]: remove this var.

File: lib/rules/disallow-newline-before-block-statements.js [hunk 10]
```diff
@@ -90,15 +184,82 @@ module.exports.prototype = {
...
+            break;
+        case 'CatchClause':
+            blockType = 'catch';
+            break;
+        default:
+            blockType = 'other';
+            break;
     }
-};
+    return blockType;
```
Comment (mikesherov) [inline lib/rules/disallow-newline-before-block-statements.js]: remove.

File: lib/rules/disallow-newline-before-block-statements.js [hunk 11]
```diff
@@ -90,15 +184,82 @@ module.exports.prototype = {
...
-        file.iterateNodesByType('BlockStatement', function(node) {
+        if (this._setting === true) {
+            checkAll(file, errors);
+        } else {
+            checkSpecificTypes(file, errors, this._setting);
+        }
+    }
+};
+
+function checkAll(file, errors) {
```
Comment (mikesherov) [inline lib/rules/disallow-newline-before-block-statements.js]: no need for this function, you can consolidate this into the checkSpecificTypes function.

File: lib/rules/disallow-newline-before-block-statements.js [hunk 12]
```diff
@@ -90,15 +184,82 @@ module.exports.prototype = {
     },
 
     check: function(file, errors) {
-        file.iterateNodesByType('BlockStatement', function(node) {
+        if (this._setting === true) {
```
Comment (mikesherov) [inline lib/rules/disallow-newline-before-block-statements.js]: again, no reason for separate functions. Consolidate them down to one check

File: lib/rules/require-newline-before-block-statements.js [hunk 13]
```diff
@@ -90,15 +182,82 @@ module.exports.prototype = {
     },
 
     check: function(file, errors) {
-        file.iterateNodesByType('BlockStatement', function(node) {
+        if (this._setting === true) {
```
Comment (mikesherov) [inline lib/rules/require-newline-before-block-statements.js]: All the comments I made on the other file apply here.

File: test/rules/disallow-newline-before-block-statements.js [hunk 14]
```diff
@@ -56,4 +56,227 @@ describe('rules/disallow-newline-before-block-statements', function() {
...
+            it('should not complain when not configured with "finally"', function() {
+                checker.configure({ disallowNewlineBeforeBlockStatements: ['for'] });
+                assert(checker.checkString('try {\n\ty++;\n} catch(e) {\n} finally\n{\n\tq = 5;\n}')
+                    .isEmpty());
+            });
+        });
+
+        describe('"while" loops', function() {
+            it('should report extra newlines when configured with "while"', function() {
+                checker.configure({ disallowNewlineBeforeBlockStatements: ['while'] });
```
Comment (mikesherov) [inline test/rules/disallow-newline-before-block-statements.js]: do this in a beforeEach for the next few tests.

File: test/rules/disallow-newline-before-block-statements.js [hunk 15]
```diff
@@ -56,4 +56,227 @@ describe('rules/disallow-newline-before-block-statements', function() {
...
+            it('should not complain when not configured with "catch"', function() {
+                checker.configure({ disallowNewlineBeforeBlockStatements: ['for'] });
+                assert(checker.checkString('try {\n\ty++;\n} catch(e)\n{\n}')
+                    .isEmpty());
+            });
+        });
+
+        describe('"finally" blocks', function() {
+            it('should report extra newlines when configured with "finally"', function() {
+                checker.configure({ disallowNewlineBeforeBlockStatements: ['finally'] });
```
Comment (mikesherov) [inline test/rules/disallow-newline-before-block-statements.js]: do this in a beforeEach for this describe.

File: test/rules/disallow-newline-before-block-statements.js [hunk 16]
```diff
@@ -56,4 +56,227 @@ describe('rules/disallow-newline-before-block-statements', function() {
...
+            it('should not complain when not configured with "try"', function() {
+                checker.configure({ disallowNewlineBeforeBlockStatements: ['for'] });
+                assert(checker.checkString('try\n{\n\ty++;\n} catch(e) {\n}')
+                    .isEmpty());
+            });
+        });
+
+        describe('"catch" blocks', function() {
+            it('should report extra newlines when configured with "catch"', function() {
+                checker.configure({ disallowNewlineBeforeBlockStatements: ['catch'] });
```
Comment (mikesherov) [inline test/rules/disallow-newline-before-block-statements.js]: do this in a beforeEach for this describe.

File: test/rules/disallow-newline-before-block-statements.js [hunk 17]
```diff
@@ -56,4 +56,227 @@ describe('rules/disallow-newline-before-block-statements', function() {
...
+            it('should not complain when not configured with "function"', function() {
+                checker.configure({ disallowNewlineBeforeBlockStatements: ['if'] });
+                assert(checker.checkString('var z = function(y)\n{\n\ty++;\n}')
+                    .isEmpty());
+            });
+        });
+
+        describe('"try" blocks', function() {
+            it('should report extra newlines when configured with "try"', function() {
+                checker.configure({ disallowNewlineBeforeBlockStatements: ['try'] });
```
Comment (mikesherov) [inline test/rules/disallow-newline-before-block-statements.js]: do this in a beforeEach for this describe.

File: test/rules/disallow-newline-before-block-statements.js [hunk 18]
```diff
@@ -56,4 +56,227 @@ describe('rules/disallow-newline-before-block-statements', function() {
...
+            it('should not complain when not configured with "function"', function() {
+                checker.configure({ disallowNewlineBeforeBlockStatements: ['if'] });
+                assert(checker.checkString('function myFunc(y)\n{\n\ty++;\n}')
+                    .isEmpty());
+            });
+        });
+
+        describe('function statements', function() {
+            it('should report extra newlines when configured with "function"', function() {
+                checker.configure({ disallowNewlineBeforeBlockStatements: ['function'] });
```
Comment (mikesherov) [inline test/rules/disallow-newline-before-block-statements.js]: do this in a beforeEach for this describe.

File: test/rules/disallow-newline-before-block-statements.js [hunk 19]
```diff
@@ -56,4 +56,227 @@ describe('rules/disallow-newline-before-block-statements', function() {
...
+            it('should not complain when not configured with "for"', function() {
+                checker.configure({ disallowNewlineBeforeBlockStatements: ['if'] });
+                assert(checker.checkString('for (var i in x)\n{\n\ty++;\n}')
+                    .isEmpty());
+            });
+        });
+
+        describe('function declarations', function() {
+            it('should report extra newlines when configured with "function"', function() {
+                checker.configure({ disallowNewlineBeforeBlockStatements: ['function'] });
```
Comment (mikesherov) [inline test/rules/disallow-newline-before-block-statements.js]: do this in a beforeEach for this describe.

File: test/rules/disallow-newline-before-block-statements.js [hunk 20]
```diff
@@ -56,4 +56,227 @@ describe('rules/disallow-newline-before-block-statements', function() {
...
+            it('should not complain when note configured with "for"', function() {
+                checker.configure({ disallowNewlineBeforeBlockStatements: ['if'] });
+                assert(checker.checkString('for (var i = 0, len = 10; i < 10; ++i)\n{\n\tx++;\n}')
+                    .isEmpty());
+            });
+        });
+
+        describe('"for...in" loops', function() {
+            it('should report extra newlines when configured with "for"', function() {
+                checker.configure({ disallowNewlineBeforeBlockStatements: ['for'] });
```
Comment (mikesherov) [inline test/rules/disallow-newline-before-block-statements.js]: do this in a beforeEach for this describe.

File: test/rules/disallow-newline-before-block-statements.js [hunk 21]
```diff
@@ -56,4 +56,227 @@ describe('rules/disallow-newline-before-block-statements', function() {
...
+            it('should not complain when not configured with "else"', function() {
+                checker.configure({ disallowNewlineBeforeBlockStatements: ['if'] });
+                assert(checker.checkString('if(i == 0) {\n\tx++;\n\ty = {\n\t\tb: "1"\n\t};\n}\nelse\n{\n\tx--;\n}')
+                    .isEmpty());
+            });
+        });
+
+        describe('"for" loops', function() {
+            it('should report extra newlines when configured with "for"', function() {
+                checker.configure({ disallowNewlineBeforeBlockStatements: ['for'] });
```
Comment (mikesherov) [inline test/rules/disallow-newline-before-block-statements.js]: do this in a beforeEach for this describe.

File: test/rules/disallow-newline-before-block-statements.js [hunk 22]
```diff
@@ -56,4 +56,227 @@ describe('rules/disallow-newline-before-block-statements', function() {
...
+            it('should not complain when not configured with "if"', function() {
+                checker.configure({ disallowNewlineBeforeBlockStatements: ['else'] });
+                assert(checker.checkString('if(i == 0)\n{\n\tx++;\n\ty = {\n\t\tb: "1"\n\t};\n}')
+                    .isEmpty());
+            });
+        });
+
+        describe('"else" and "else if" blocks', function() {
+            it('should report extra newlines when configured with "else"', function() {
+                checker.configure({ disallowNewlineBeforeBlockStatements: ['else'] });
```
Comment (mikesherov) [inline test/rules/disallow-newline-before-block-statements.js]: do this in a beforeEach for this describe.

File: test/rules/disallow-newline-before-block-statements.js [hunk 23]
```diff
@@ -56,4 +56,227 @@ describe('rules/disallow-newline-before-block-statements', function() {
             assert(checker.checkString('{ test: 1 }').isEmpty());
         });
     });
+
+    describe('with option value array - ', function() {
+
+        describe('"if" blocks', function() {
+            it('should report extra newlines when configured with "if"', function() {
+                checker.configure({ disallowNewlineBeforeBlockStatements: ['if'] });
```
Comment (mikesherov) [inline test/rules/disallow-newline-before-block-statements.js]: do this in a beforeEach for this describe.

File: test/rules/disallow-newline-before-block-statements.js [hunk 24]
```diff
@@ -56,4 +56,227 @@ describe('rules/disallow-newline-before-block-statements', function() {
             assert(checker.checkString('{ test: 1 }').isEmpty());
         });
     });
+
+    describe('with option value array - ', function() {
+
```
Comment (mikesherov) [inline test/rules/disallow-newline-before-block-statements.js]: remove newline please.

File: test/rules/disallow-newline-before-block-statements.js [hunk 25]
```diff
@@ -56,4 +56,227 @@ describe('rules/disallow-newline-before-block-statements', function() {
...
+            it('should not complain when not configured with "while"', function() {
+                checker.configure({ disallowNewlineBeforeBlockStatements: ['for'] });
+                assert(checker.checkString('while (x < 10)\n{\n\tx++;\n}')
+                    .isEmpty());
+            });
+        });
+
+        describe('"do...while" loops', function() {
+            it('should report extra newlines when configured with "do"', function() {
+                checker.configure({ disallowNewlineBeforeBlockStatements: ['do'] });
```
Comment (mikesherov) [inline test/rules/disallow-newline-before-block-statements.js]: do this in a beforeEach for this describe.

File: test/rules/require-newline-before-block-statements.js [hunk 26]
```diff
@@ -54,4 +54,227 @@ describe('rules/require-newline-before-block-statements', function() {
             assert(checker.checkString('{test: 1 }').isEmpty());
         });
     });
+
+    describe('with option value array - ', function() {
+
```
Comment (mikesherov) [inline test/rules/require-newline-before-block-statements.js]: all comments from previous file apply to this file as well.

**Comment by mikesherov** [comment]
@davejhilton thanks for contributing! I made a lot of comments, but they're minor, so... good work.

Is there an existing issue for this behavior? If so, please reference it in your commit message.

Also, it looks like you aren't catching all cases of the code you wrote. Try running `npm run-script coverage-html && open coverage.html` from the command line to look at the code coverage report and see what you missed.

Thanks again and let me know if you have any questions!

**Comment by davejhilton** [comment]
Thanks,

I've now fixed the stylistic and code-coverage concerns mentioned above.

File: lib/rules/disallow-newline-before-block-statements.js [hunk 27]
```diff
@@ -90,15 +183,57 @@ module.exports.prototype = {
     },
 
     check: function(file, errors) {
-        file.iterateNodesByType('BlockStatement', function(node) {
-            var openingBrace = file.getFirstNodeToken(node);
-            var prevToken = file.getPrevToken(openingBrace);
+        var setting = this._setting;
+        if (setting === true) {
+            file.iterateNodesByType('BlockStatement', function(node) {
```
Comment (mikesherov) [inline lib/rules/disallow-newline-before-block-statements.js]: still can consolidate this into one iteration over BlockStatement, no reason to split the two.

File: lib/rules/disallow-newline-before-block-statements.js [hunk 28]
```diff
@@ -90,15 +183,57 @@ module.exports.prototype = {
...
+                    token: prevToken,
+                    nextToken: openingBrace,
+                    message: 'Newline before curly brace for block statement is disallowed'
+                });
             });
-        });
+        } else {
+            file.iterateNodesByType(['BlockStatement'], function(node) {
+                var type = getBlockType(node);
+                if (setting.indexOf(type) !== -1) {
```
Comment (mikesherov) [inline lib/rules/disallow-newline-before-block-statements.js]: `if (setting === true || setting.indexOf(type) !== -1) {` and then the outer if/else goes away.

**Comment by mikesherov** [comment]
Thanks again! Looks good now.

**Comment by davejhilton** [comment]
Awesome. So what's left before it gets merged?

**Comment by mikesherov** [comment]
It'll get merged soon. 1.11 was just released so I want to give it a day or two to see if anything needs to go into a 1.11.1 first.

**Comment by davejhilton** [comment]
Sounds good, thanks!

**Comment by mikesherov** [comment]
No, thank you!

**Comment by TheSavior** [comment]
Somewhat unrelated, but I don't think I agree with the name of this rule. It seems like it should be something like (require|disallow)NewlineBeforeOpeningCurly since it isn't a newline before the block.

**Comment by mikesherov** [comment]
@TheSavior can't be `OpeningCurly` because it doesn't deal with object literals nor destructuring assignments.

**Comment by markelog** [comment]
Thank you for the patience! And sorry it took that long to land.

**Pull request closed without merging**